src/requestManager.cpp

```diff
@@ -435,6 +436,73 @@ bool RequestBlock(CNode *pfrom, CInv obj)
                 return true;
             }
         }
+        // BUIP010 Xtreme Thinblocks: end section
+        // BUIPXXX Graphene blocks: begin section
+        else if (IsGrapheneBlockEnabled() && IsChainNearlySyncd())
+        {
+            if (HaveConnectGrapheneNodes() || (HaveGrapheneNodes() && graphenedata.CheckGrapheneBlockTimer(obj.hash)))
+            {
+                // Must download a graphene block from a graphene enabled peer.
+                // We can only request one graphene block per peer at a time.
+                if (pfrom->mapGrapheneBlocksInFlight.size() < 1 && CanGrapheneBlockBeDownloaded(pfrom))
+                {
+                    AddGrapheneBlockInFlight(pfrom, inv2.hash);
+
+                    inv2.type = MSG_GRAPHENEBLOCK;
+                    std::vector<uint256> vOrphanHashes;
+                    {
+                        LOCK(cs_orphancache);
+                        for (map<uint256, COrphanTx>::iterator mi = mapOrphanTransactions.begin();
+                             mi != mapOrphanTransactions.end(); ++mi)
+                            vOrphanHashes.push_back((*mi).first);
+                    }
+                    // Instead of building a bloom filter here, we actually just need to fill in CMempoolInfo
+                    ss << inv2;
+                    ss << GetGrapheneMempoolInfo();
+                    MarkBlockAsInFlight(pfrom->GetId(), obj.hash, chainParams.GetConsensus());
+                    pfrom->PushMessage(NetMsgType::GET_GRAPHENE, ss);
+                    LOG(GRAPHENE, "Requesting graphene block %s from peer %s (%d)\n", inv2.hash.ToString(),
+                        pfrom->addrName.c_str(), pfrom->id);
+                    return true;
+                }
+            }
+            else
+            {
+                // Try to download a graphene block if possible otherwise just download a regular block.
+                // We can only request one graphene block per peer at a time.
+                MarkBlockAsInFlight(pfrom->GetId(), obj.hash, chainParams.GetConsensus());
+                if (pfrom->mapGrapheneBlocksInFlight.size() < 1 && CanGrapheneBlockBeDownloaded(pfrom))
+                {
+                    AddGrapheneBlockInFlight(pfrom, inv2.hash);
+
+                    inv2.type = MSG_GRAPHENEBLOCK;
+                    std::vector<uint256> vOrphanHashes;
+                    {
+                        LOCK(cs_orphancache);
+                        for (map<uint256, COrphanTx>::iterator mi = mapOrphanTransactions.begin();
+                             mi != mapOrphanTransactions.end(); ++mi)
+                            vOrphanHashes.push_back((*mi).first);
+                    }
+                    BuildSeededBloomFilter(filterMemPool, vOrphanHashes, inv2.hash, pfrom);
```

Comment:
Why create a seeded bloom filter here?  this is what we need for an Xthin.  It looks like you're possibly not  creating a proper graphene block here?

Reply:
Whoops. That's a copy paste error. I've removed that call to BuildSeededBloomFilter and added a line like in 461.